rules/internal/utils/http.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"github.com/golang/protobuf/proto"
+	"github.com/jhump/protoreflect/desc"
+	"google.golang.org/genproto/googleapis/api/annotations"
+)
+
+// GetHTTPRules returns a slice of HTTP rules for a given method descriptor.
+//
+// Note: This returns a slice -- it takes the google.api.http annotation,
+// and then flattens the values in `additional_bindings`.
+// This allows rule authors to simply range over all of the HTTP rules,
+// since the common case is to want to apply the checks to all of them.
+func GetHTTPRules(m *desc.MethodDescriptor) []*annotations.HttpRule {
+	rules := []*annotations.HttpRule{}
+
+	// Get the method options.
+	opts := m.GetMethodOptions()
+
+	// Get the "primary" rule (the direct google.api.http annotation).
+	var httpRule *annotations.HttpRule
+	if x, err := proto.GetExtension(opts, annotations.E_Http); err == nil {
+		httpRule = x.(*annotations.HttpRule)
+		rules = append(rules, httpRule)
+	}
+
+	// Add any additional bindings and flatten them into `rules`.
+	rules = append(rules, httpRule.GetAdditionalBindings()...)
```

Comment:
I'd suggest moving this into the `if` block, remove line 36, and change 38 to `:=`

You can also optionally change line 30 to `var rules []*annotations.HttpRule` to avoid an unnecessary allocation (I believe this is a matter of preference because I've seen both forms suggested in readability reviews). The zero value of slices is intended to be useful by itself (i.e. you can call `append` and `len` on it, which will perform allocations if needed).

Reply:
Done. I chose to leave line 30 as-is.